File: leetcode/javascript/50-myPow.js
/**
 * 实现 pow(x, n) ，即计算 x 的 n 次幂函数（即，xn）。
 */

/**
 * @description Pow(x, n)
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
	var compute = function(num, n) {
		if (n === 0) {
			return 1;
		}

		if (n === 1) {
			return num;
		}

		var res = compute(num, Math.floor(n / 2));

		if (n % 2) {
			return res * res * num;
		}

		return res * res;
	}

	return n > 0 ? compute(x, n).toFixed(5) : (1 / compute(x, -n)).toFixed(5);
}

/**
 * 测试用例：
 * console.log(myPow(2, 10));
 */

/**
 * 本题核心： 递归、二分法
 *
 * 反思：注意在递归时考虑能否进行尾递归优化
 */

File: leetcode/python/50-myPow.py
from math import floor

class Solution:
    def myPow(self, x, n):
        def formatResult(res):
            return float('%.5f' % res)
            
        def compute(num, n):
            if (n == 0):
                return 1
            if (n == 1):
                return num

            res = compute(x, floor(n / 2))

            if (n % 2):
                return res * res * num
            else:
                return res * res

        res = 0
        if (n <= 0):
            res = 1.0 / compute(x, -n)
        else:
            res = compute(x, n)

        return formatResult(res)

# solution = Solution()
# print(solution.myPow(2, -2))
